src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java

```diff
@@ -290,10 +290,11 @@ private boolean runRepair() {
       }
 
       try (Timer.Context cxt1 = context.metricRegistry.timer(metricNameForRepairing(segment)).time()) {
+        boolean segmentsLocked = false;
         try {
           LOG.debug("Enter synchronized section with segment ID {}", segmentId);
           synchronized (condition) {
-            if (!lockSegmentRunners()) {
+            if (!(segmentsLocked = lockSegmentRunners())) {
```

Comment:
This will conflict with a fix I have in #732 : https://github.com/thelastpickle/cassandra-reaper/pull/732/commits/29291f15c9ac11b3995e1f5995cca7d39e03b1b6
Also, this patch is incomplete as we're still releasing the lock twice when a segment is successfully triggered.

I tried to do it the way you did, but abandoned because the `segmentsLocked` boolean could get set to false by a concurrent thread since it's outside the synchronized section.

For what it's worth, I've tested my fix on several attempts with up to 9 nodes, after getting into stuck repairs. They all passed nicely after that and I had no more errors.

Reply:
> I tried to do it the way you did, but abandoned because the segmentsLocked boolean could get set to false by a concurrent thread since it's outside the synchronized section.

No it can't. It's a variable to the method, a separate thread will have its own `segmentsLocked` variable.

> Also, this patch is incomplete as we're still releasing the lock twice when a segment is successfully triggered.

Fixed 👍 
Added the line `segmentsLocked = false;` in
```
            if (0 != repairNo) {
              releaseSegmentRunners();
              segmentsLocked = false;
              processTriggeredSegment(segment, coordinator, repairNo);
            } else {
```